### DISCUSSION TITLE
Announcement: Concerning the `use-wayland` feature

### DISCUSSION THREAD

**Discussion by MoAlyousef** [fltk-rs/fltk-rs#1281] [General]
Hello

- The next release of fltk-rs will be pulling the updates from FLTK upstream, which has made the option OPTION_USE_WAYLAND generate a hybrid FLTK application, capable of running on wayland (using Cairo for drawing), and also capable of running on X11 systems without wayland support. This makes FLTK apps comparable to GTK apps in this regard.

-  The wayland backend can be disabled by running the app with the FLTK_BACKEND=x11 env variable:
```
FLTK_BACKEND=x11 myapp # assuming the app is globally installed
```
The app can also be wrapped in a shell script that's globally exposed which would launch the app with the FLTK_BACKEND set.

- For dev purposes, you can also check your app for any discrepencies in appearance or behavior by launching with FLTK_BACKEND=x11 and FLTK_BACKEND=wayland.

- You can also disable the wayland backend programmatically using `app::disable_wayland();` function.

Regarding CI builds, the `use-wayland` feature will require the CI runners to install xlib development headers and libs whereas previously this wasn't required. For example the required dev dependencies for a debian-based distro (like the gha linux runner):
```
sudo apt-get install -y libx11-dev libxext-dev libxft-dev libxinerama-dev libxcursor-dev libxrender-dev libxfixes-dev libxkbcommon-dev libwayland-dev wayland-protocols libpango1.0-dev libdbus-1-dev
```

**Comment by MoAlyousef** [comment]
If only wayland is needed (without X11), you can set the CFLTK_WAYLAND_ONLY=1 environment variable prior to running the build.

**Comment by adigitoleo** [comment]
I have been unsuccessful in building a hello world example with fltk-rs 1.3.16 using `CFLTK_WAYLAND_ONLY=1 cargo build`. During linking to libX11 I get an error because of an undefined references. This happened with or without the `fltk-bundled` feature and on two different machines. Both have libX11 1.8.1 including headers.

I am able to build without that env variable but when I launch the app it always uses XWayland, and FLTK_BACKEND does not seem to have any effect. Can you confirm that the hello world examples in the docs should still work with wayland? Are there other things I can check?

I noticed that FLTK provides a `--disable-x11` configure flag, I wonder if fltk-rs can (or already does?) inject that into the bundled fltk compilation somehow, or maybe it's an issue with my environment.

**Comment by MoAlyousef** [comment]
The available bundles aren’t built against wayland, so that call alone would fail.  It would have to be:
```
CFLTK_WAYLAND_ONLY=1 cargo build —features=fltk/use-wayland
```
Or you can set the feature in your Cargo.toml.

**Reply by adigitoleo** [reply]
Ah, I should have clarified, I get the error despite setting the `use-wayland` feature in Cargo.toml.

**Reply by MoAlyousef** [reply]
Can you post the error?
Also what wayland compositor are you using?

Locally I’m able to build and run with both use-wayland hybrid and only wayland variants. I can know it’s not using XWaylabd because of the different window decorations and some minor drawing differences. Also for some reason the wayland window is a bit smaller.

**Reply by adigitoleo** [reply]
I'm building on a machine that has sway, although right now I'm only in a remote session so I can't test the app until tomorrow.

```
  = note: /usr/bin/ld: /home/admin/issuedrafts/fltk-rs-wayland/eg-fltk-wayland/target/debug/deps/libfltk_sys-760e5c0dc7c3fd1b.rlib(Fl_X11_Screen_Driver.cxx.o): undefined reference to symbol 'XGetImage'
          /usr/bin/ld: /usr/lib/libX11.so.6: error adding symbols: DSO missing from command line
          collect2: error: ld returned 1 exit status

  = help: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)

error: could not compile `eg-fltk-wayland` due to previous error
```

**Reply by MoAlyousef** [reply]
Can you check also this repo for example:
https://github.com/MoAlyousef/test_wayland/actions/runs/3256076536

It builds a hybrid and pure wayland fltk app.

**Reply by MoAlyousef** [reply]
If you build an app purely for wayland on an X11 environment. It will still use libcairo from your system, but that libcairo is built against xlib, so it will still pull its x11 dependencies.

**Reply by adigitoleo** [reply]
Thanks for the help, unfortunately I still have no luck building the linked example repo with:

`CFLTK_WAYLAND_ONLY=1 cargo build --features=fltk/use-wayland`

The build fails with the same libX11 `undefined reference to symbol ...` error. Seems like there is some kind of problem because of an incompatible X11-related lib version... I have all of the X11 deps that you listed, but I'm not on Ubuntu:

```
libX11-devel-1.8.1_1
libXext-devel-1.3.4_1
libXft-devel-2.3.6_1
libXinerama-devel-1.1.4_1
libXcursor-devel-1.2.1_1
libXrender-devel-0.9.10_2
libXfixes-devel-6.0.0_1
libxkbcommon-devel-1.4.1_1
wayland-protocols-1.26_1
pango-devel-1.50.10_2
dbus-devel-1.14.2_1
cairo-devel-1.16.0_2
```

When I get time I'll try building upstream fltk and running a simple C example to see if that works.

**Reply by MoAlyousef** [reply]
Can you also post the output from:
```
cargo clean && CFLTK_WAYLAND_ONLY=1 cargo build --features=fltk/use-wayland -vv
```
This should show tell us how FLTK is being configured on your system.

**Reply by adigitoleo** [reply]
The build log is here: https://x0.at/tsNP.txt

**Reply by MoAlyousef** [reply]
The build script doesn’t find cmake or a c++ compiler, so it defaults to pulling the prebuilt binaries which are only available for x11.

**Reply by adigitoleo** [reply]
Whoops, tried again with cmake installed but got this: https://x0.at/U_bN.txt

I guess the next option is to compile fltk and link to it directly. I've cloned and compiled the fltk sources with the same flags as your CI example: https://github.com/MoAlyousef/test_wayland/blob/main/.github/workflows/rust.yml

Could you help me out with how I tell cargo/cmake to use that one?

**Reply by MoAlyousef** [reply]
The output shows that the fltk-bundled flag is still being passed to your build. What's the content of your Cargo.toml?

To build against a locally built version of FLTK, it should be installed in a system path (i.e you need to run (sudo cmake --build bin --target install).
Then you would use the system-fltk feature, in addition to the use-wayland feature.

**Comment by MoAlyousef** [comment]
FLTK upstream now enable the wayland hybrid backend by default. This is currently not the case for fltk-rs, but it may be changed in a future release.
